src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -676,61 +677,72 @@ public long GetNumberOfCounterEntries(DocumentsOperationContext context)
             return fst.NumberOfEntries;
         }
 
-        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId)
+        public void UpdateDocumentCounters(DocumentsOperationContext context, BlittableJsonReaderObject doc, string docId,
+            SortedSet<string> countersToAdd, HashSet<string> countersToRemove)
         {
+            if (countersToRemove.Count == 0 && countersToAdd.Count == 0)
+                return;
+
             BlittableJsonReaderArray metadataCounters = null;
             if (doc.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata))
             {
                 metadata.TryGet(Constants.Documents.Metadata.Counters, out metadataCounters);
             }
 
-            var counters = GetCountersForDocument(context, docId).ToList();
+            var counters = GetCountersForDocument(metadataCounters, countersToAdd, countersToRemove, out var hadModifications);
+            if (hadModifications == false)
+                return;
 
-            var flags = DocumentFlags.None;
-            if (counters.Count == 0)
+            var flags = counters.Count == 0 ? DocumentFlags.None : DocumentFlags.HasCounters;
+            doc.Modifications = new DynamicJsonValue(doc);
+            if (metadata == null)
             {
-                if (metadataCounters == null)
+                doc.Modifications[Constants.Documents.Metadata.Key] = new DynamicJsonValue
                 {
-                    return;
-                }
-
-                metadata.Modifications = new DynamicJsonValue(metadata);
-                metadata.Modifications.Remove(Constants.Documents.Metadata.Counters);
```

Comment:
I think you missed `Remove(Constants.Documents.Metadata.Counters)` if `counters.Count == 0 `

Reply:
Right. Done.

Reply:
Reminder - no fix was pushed